include/cbor/stl.h

```diff
@@ -571,6 +572,76 @@ struct traits<std::map<KeyType, ValueType>>
   }
 };
 
+/**
+ * Specialization for std::shared_ptr.
+ */
+template <typename PointedType>
+struct traits<std::shared_ptr<PointedType>>
+{
+  using Type = std::shared_ptr<PointedType>;
+  template <typename Data>
+  static result serializer(const std::shared_ptr<PointedType>& v, Data& data)
+  {
+    if (v == nullptr)
+    {
+      return to_cbor(nullptr, data);
+    }
+    return to_cbor(*v, data);
+  }
+
+  template <typename Data>
+  static result deserializer(Type& v, Data& data)
+  {
+    if (data.isNull())
+    {
+      v = nullptr;
+      return data.advance(1);
+    }
+    else
+    {
+      v = std::make_shared<PointedType>();
+      auto res = from_cbor(*v, data);
+      return res;
+    }
+    return false;
+  }
+};
+
+/**
+ * Specialization for std::unique_ptr.
+ */
+template <typename PointedType>
```

Comment:
The code for `std::shared_ptr` and `std::unique_ptr` is the same, so I wonder if you could use `template template parameters` :thinking: 

You probably need a helper `struct` and instead of `std::make_` using `.reset(new`.

Likely not worth it.

Reply:
No, I considered it, but then you do indeed need another trait system for the `make_new` and `make_unique` :(

Reply:
If that's the only issue, why don't you use `.reset(new`? which has the added benefit that doesn't call the global `operator new` if it's overloaded by the `PointedType`, for `std::shared_ptr`.

See https://en.cppreference.com/w/cpp/memory/shared_ptr/make_shared

Reply:
Leaving as is, I'd have to introduce a new family to the trait system and this helper struct. This would definitely be just as much code, and less readable than this direct insertion into the trait system.

Reply:
Fair enough